src/core/client_channel/load_balanced_call_destination.cc

```diff
@@ -0,0 +1,334 @@
+    auto* legacy_context = GetContext<grpc_call_context_element>();
+    return static_cast<ClientCallTracer::CallAttemptTracer*>(
+        legacy_context[GRPC_CONTEXT_CALL_TRACER].value);
+  }
+};
+
+// TODO(roth): Remove this in favor of the gprpp Match() function once
+// we can do that without breaking lock annotations.
+template <typename T>
+T HandlePickResult(
+    LoadBalancingPolicy::PickResult* result,
+    std::function<T(LoadBalancingPolicy::PickResult::Complete*)> complete_func,
+    std::function<T(LoadBalancingPolicy::PickResult::Queue*)> queue_func,
+    std::function<T(LoadBalancingPolicy::PickResult::Fail*)> fail_func,
+    std::function<T(LoadBalancingPolicy::PickResult::Drop*)> drop_func) {
+  auto* complete_pick =
+      absl::get_if<LoadBalancingPolicy::PickResult::Complete>(&result->result);
+  if (complete_pick != nullptr) {
+    return complete_func(complete_pick);
+  }
+  auto* queue_pick =
+      absl::get_if<LoadBalancingPolicy::PickResult::Queue>(&result->result);
+  if (queue_pick != nullptr) {
+    return queue_func(queue_pick);
+  }
+  auto* fail_pick =
+      absl::get_if<LoadBalancingPolicy::PickResult::Fail>(&result->result);
+  if (fail_pick != nullptr) {
+    return fail_func(fail_pick);
+  }
+  auto* drop_pick =
+      absl::get_if<LoadBalancingPolicy::PickResult::Drop>(&result->result);
+  CHECK(drop_pick != nullptr);
+  return drop_func(drop_pick);
+}
+
+// Does an LB pick for a call.  Returns one of the following things:
+// - Continue{}, meaning to queue the pick
+// - a non-OK status, meaning to fail the call
+// - a connected subchannel, meaning that the pick is complete
+// When the pick is complete, pushes client_initial_metadata onto
+// call_initiator.  Also adds the subchannel call tracker (if any) to
+// context.
+LoopCtl<absl::StatusOr<RefCountedPtr<UnstartedCallDestination>>> PickSubchannel(
+    LoadBalancingPolicy::SubchannelPicker& picker,
+    UnstartedCallHandler& unstarted_handler) {
+  // Perform LB pick.
+  auto& client_initial_metadata =
+      unstarted_handler.UnprocessedClientInitialMetadata();
+  LoadBalancingPolicy::PickArgs pick_args;
+  Slice* path = client_initial_metadata.get_pointer(HttpPathMetadata());
+  CHECK(path != nullptr);
+  pick_args.path = path->as_string_view();
+  LbCallState lb_call_state;
+  pick_args.call_state = &lb_call_state;
+  LbMetadata initial_metadata(&client_initial_metadata);
+  pick_args.initial_metadata = &initial_metadata;
+  auto result = picker.Pick(pick_args);
+  // Handle result.
+  return HandlePickResult<
+      LoopCtl<absl::StatusOr<RefCountedPtr<UnstartedCallDestination>>>>(
+      &result,
+      // CompletePick
+      [&](LoadBalancingPolicy::PickResult::Complete* complete_pick)
+          -> LoopCtl<absl::StatusOr<RefCountedPtr<UnstartedCallDestination>>> {
+        if (GRPC_TRACE_FLAG_ENABLED(grpc_client_channel_lb_call_trace)) {
+          gpr_log(GPR_INFO,
+                  "client_channel: %sLB pick succeeded: subchannel=%p",
```

Comment:
The fact that we're no longer logging the address of the `ClientChannel` is going to make debugging harder whenever there are multiple channels in the application, which is quite common.  Can we somehow plumb that address back in here, so that we can more easily differentiate between channels?

Reply:
Soon(TM). And let's not block on it.

I want to build a type that, for a given type of object (we'll have multiple instances of these) tracks the identity of the originating object, without necessarily keeping that object in memory - so we don't couple object lifetimes just for the sake of logging... so in ClientChannel you'd have a member `ChannelLogId id_;` and pass that object to all the child objects we create. When we log we'd `LOG(INFO) << id_ << "blah blah";` and expect a consistent prefix to be added.

I have some ideas on how to do this simply, efficiently, and type-safely -- but I want @drfloob to land #36576 first - the codegen there is going to make this a little easier.